dockstore-webservice/src/main/java/io/dockstore/webservice/languages/WDLHandler.java

```diff
@@ -579,11 +577,50 @@ public static boolean versionIsGreaterThanCurrentlySupported(String semVerString
 
     /**
      * Get the version from the WDL descriptor file content. If no version is found, return the default WDL version defined by the WDL spec.
-     * @param descriptorContent
+     * If the version is invalid, return Optional.empty()
+     *
+     * @param primaryDescriptorPath The absolute path of the descriptor SourceFile to get the 'version' from
+     * @param sourceFiles           A set of SourceFiles containing the primary descriptor SourceFile and any imports
      * @return
      */
-    public static String getLanguageVersion(String descriptorContent) {
-        return getSemanticVersionString(descriptorContent).orElse(DEFAULT_WDL_VERSION);
+    public static Optional<String> getLanguageVersion(String primaryDescriptorPath, Set<SourceFile> sourceFiles) {
+        WdlBridge wdlBridge = new WdlBridge();
+
+        Optional<SourceFile> primaryDescriptor = sourceFiles.stream().filter(sourceFile -> sourceFile.getAbsolutePath().equals(primaryDescriptorPath)).findFirst();
+        if (primaryDescriptor.isEmpty()) {
+            return Optional.empty();
+        }
+
+        final String primaryDescriptorContent = primaryDescriptor.get().getContent();
+        final Map<String, String> secondaryFiles = sourceFiles.stream()
+                .filter(sourceFile -> !sourceFile.getAbsolutePath().equals(primaryDescriptorPath))
+                .collect(Collectors.toMap(SourceFile::getAbsolutePath, SourceFile::getContent));
+        wdlBridge.setSecondaryFiles((HashMap<String, String>)secondaryFiles);
```

Comment:
Casting the `Map` to a `HashMap` is in theory fragile -- according to the `Collectors.toMap` JavaDoc: "There are no guarantees on the type, mutability, serializability, or thread-safety of the {@code Map} returned". On the other hand, it does return a `HashMap` today, and probably always will.

To be paranoid, you could either create a new HashMap from the Map, or change the signature of `setSecondaryFiles` to take a `Map`, which is probably how it should have been written in the first place. I'd suggest just creating a new HashMap; changing the signature seems like a bit more work.

Reply:
The current implementation of `Collectors.toMap`, in OpenJDK:
https://github.com/openjdk/jdk/blob/f729f5b6d01b0d3a0ee21f50199ca30935c8237a/src/java.base/share/classes/java/util/stream/Collectors.java#L1435
Yeah, right now, it appears to always returns `HashMap`.  But, as Charles mentions, it's entirely possible that someday it could return something else (immutable empty or single-entry Maps being a likely candidate)...

Reply:
I created a new HashMap. Still had to cast it though, but we know it's a HashMap

Reply:
FWIW, you can probably define `secondaryFiles` as a `HashMap` and eliminate the cast.

Reply:
Weird, the old code is showing up, my comment doesn't apply to it. :)